src/libsystemd/sd-json/json-util.c

```diff
@@ -281,3 +284,124 @@ int json_dispatch_pidref(const char *name, sd_json_variant *variant, sd_json_dis
+                return sd_json_variant_new_null(ret);
+
+        return sd_json_buildo(
+                        ret,
+                        SD_JSON_BUILD_PAIR_UNSIGNED("major", major(devnum)),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("minor", minor(devnum)));
+}
+
+int json_dispatch_devnum(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        dev_t *ret = ASSERT_PTR(userdata);
+        int r;
+
+        assert(variant);
+
+        if (sd_json_variant_is_null(variant)) {
+                *ret = 0;
+                return 0;
+        }
+
+        struct {
+                uint32_t major;
+                uint32_t minor;
+        } data;
+
+        static const sd_json_dispatch_field dispatch_table[] = {
+                { "major", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint32, voffsetof(data, major), SD_JSON_MANDATORY },
+                { "minor", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint32, voffsetof(data, minor), SD_JSON_MANDATORY },
+                {}
+        };
+
+        r = sd_json_dispatch(variant, dispatch_table, flags, &data);
+        if (r < 0)
+                return r;
+
+        if (!DEVICE_MAJOR_VALID(data.major) || !DEVICE_MINOR_VALID(data.minor))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not a valid device number.", strna(name));
+
+        *ret = makedev(data.major, data.minor);
+        return 0;
+}
+
+static int json_variant_new_stat(sd_json_variant **ret, const struct stat *st) {
+        char mode[STRLEN("0755")+1];
+
+        assert(st);
+
+        if (!stat_is_set(st))
+                return sd_json_variant_new_null(ret);
+
+        xsprintf(mode, "%04o", st->st_mode & ~S_IFMT);
+
+        return sd_json_buildo(
+                        ret,
+                        JSON_BUILD_PAIR_DEVNUM("dev", st->st_dev),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("inode", st->st_ino),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("type", inode_type_to_string(st->st_mode)),
+                        SD_JSON_BUILD_PAIR_STRING("mode", mode),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("linkCount", st->st_nlink),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("uid", st->st_uid),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("gid", st->st_gid),
+                        SD_JSON_BUILD_PAIR_CONDITION(
+                                        S_ISCHR(st->st_mode) || S_ISBLK(st->st_mode),
+                                        "rdev",
+                                        JSON_BUILD_DEVNUM(st->st_rdev)),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("size", st->st_size),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("blockSize", st->st_blksize),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("blocks", st->st_blocks));
+}
+
+static int json_variant_new_file_handle(const struct file_handle *fid, sd_json_variant **ret) {
```

Comment:
Hmm, we should still make the position ret param consistent? Place it at the beginning?